src/test/ui/const-eval/infinite_loop.stderr

```diff
@@ -0,0 +1,41 @@
+error[E0019]: constant contains unimplemented expression type
+  --> $DIR/infinite_loop.rs:20:9
+   |
+LL | /         while n != 0 { //~ ERROR constant contains unimplemented expression type
+LL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };
+LL | |         }
+   | |_________^
+
+warning: Constant evaluating a complex constant, this might take some time
+  --> $DIR/infinite_loop.rs:16:18
+   |
+LL |       let _ = [(); {
+   |  __________________^
+LL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time
+LL | |         //~| ERROR could not evaluate repeat length
+LL | |         let mut n = 113383; // #20 in A006884
+...  |
+LL | |         n
+LL | |     }];
+   | |_____^
+
+error[E0080]: could not evaluate repeat length
+  --> $DIR/infinite_loop.rs:16:18
+   |
+LL |       let _ = [(); {
+   |  __________________^
+LL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time
+LL | |         //~| ERROR could not evaluate repeat length
+LL | |         let mut n = 113383; // #20 in A006884
+LL | |         while n != 0 { //~ ERROR constant contains unimplemented expression type
+LL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };
+   | |                    ---------- program will never terminate
```

Comment:
I think it would be prudent to add more info here. Like that the highlighted expression is simply the first expression found where the constant evaluator could with certainty say that the evaluation will not terminate.

Reply:
I agree, this is a bit obtuse. What about expanding the span of the error to be the entire expression? I'll need to do some more reading into the diagnostics parts of rustc to implement that. I could also replace it with something like "Infinite loop detected here, const evaluation will never terminate." But it's hard to make clear why *this* particular statement is being singled out when the problem is the whole block.

Reply:
We can't really figure out which part is at fault. I think there's no nice solution except explain more things in the message.